Support subclass constructor capturing in export

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #147014

Notable TODOs:
1. Need to implement AutogradHOP to get rid of subclasses before serializing
2. Need to implement mechanism to figure out what subclasses will be used in export when they are not expressed in the inputs 


cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

Differential Revision: [D69640673](https://our.internmc.facebook.com/intern/diff/D69640673)

tugsbayasgalan committed:
[DRAFT] Support constructor capturing in export

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[DRAFT] Support subclass constructor capturing in export"


Notable TODOs:
1. Need to implement support capturing container, non fx-able type capture in constructor 
2. Need to implement AutogradHOP to get rid of subclasses before serializing
3. Need to implement mechanism to figure out what subclasses will be used in export when they are not expressed in the inputs 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

tugsbayasgalan commented on torch/_export/wrappers.py:
```diff
@@ -115,3 +125,47 @@ def call(self, *args):
+                    
+                    flat_args, in_spec = to_graphable((tuple(args[1:]), kwargs))
+
+                    qualname = tracer.get_fresh_qualname(constructor_subclass.__name__)
```
Need better naming handling here.

tugsbayasgalan commented on torch/_export/wrappers.py:
```diff
@@ -115,3 +125,47 @@ def call(self, *args):
+                            flat_proxy_args.append(arg)
+
+                    _, func_spec = torch.utils._pytree.tree_flatten(ConstantFunction(type(subclass)))
+                    qualname = tracer.get_fresh_qualname(type(subclass).__name__)
```
Better naming

tugsbayasgalan committed:
Update on "[DRAFT] Support subclass constructor capturing in export"


Notable TODOs:
1. Need to implement AutogradHOP to get rid of subclasses before serializing
2. Need to implement mechanism to figure out what subclasses will be used in export when they are not expressed in the inputs 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[DRAFT] Support subclass constructor capturing in export"


Notable TODOs:
1. Need to implement AutogradHOP to get rid of subclasses before serializing
2. Need to implement mechanism to figure out what subclasses will be used in export when they are not expressed in the inputs 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69640673).

tugsbayasgalan committed:
Update on "[DRAFT] Support subclass constructor capturing in export"


Notable TODOs:
1. Need to implement AutogradHOP to get rid of subclasses before serializing
2. Need to implement mechanism to figure out what subclasses will be used in export when they are not expressed in the inputs 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

Differential Revision: [D69640673](https://our.internmc.facebook.com/intern/diff/D69640673)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69640673).

tugsbayasgalan committed:
Update on "[DRAFT] Support subclass constructor capturing in export"


Notable TODOs:
1. Need to implement AutogradHOP to get rid of subclasses before serializing
2. Need to implement mechanism to figure out what subclasses will be used in export when they are not expressed in the inputs 


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

Differential Revision: [D69640673](https://our.internmc.facebook.com/intern/diff/D69640673)

[ghstack-poisoned]

bdhirsh commented on torch/testing/_internal/two_tensor.py:
```diff
@@ -36,6 +37,7 @@ def __new__(cls, a, b, outer_size=None, outer_stride=None):
         assert a.storage_offset() == b.storage_offset()
         return out
 
+    @_mark_constructor_exportable_experimental
```
fyi @albanD @drisspg, new decorator that we're adding, that subclass authors need to opt into in order to support "subclass constructor inside of a graph" for export.

not sure about the name. Although to be fair, once @IvanKobzarev's "base wrapper subclass" design is further along, we can hide the constructor in there for subclass authors that use the base subclass (which will hopefully be all quantization subclasses)

albanD replied:
Since this is private API, I guess the name is not as important and we can go with as descriptive as possible.
We can always refine the name when this becomes public.

drisspg replied:
I think even in prototype state some good documentation on what exact invariants subclass author is promising to abide by would be helpful

bdhirsh commented on torch/distributed/tensor/_op_schema.py:
```diff
@@ -390,7 +390,7 @@ def gen_fake_args(self) -> ArgsType:
         """
         return tree_map_only(
-            DTensorSpec, _rebuild_tensor_from_dtensor_meta, self.args_schema
+            DTensorSpec, _rebuild_tensor_from_dtensor_meta, self.args_schema, is_leaf = lambda x: isinstance(x, DTensorSpec)
```
why do we need these?

tugsbayasgalan replied:
This is because when we register DTensorSpec as constant in pytree, we treat it as empty container hence tree_map will just ignore DTensorSpec as valid type to map over. This is consistent with how Jax.tree_map does i think. 
```
>>> stuff2 = [Point(4, 4), Point(4,4)]
>>> jax = tree_map(lambda x: x.a + 4, stuff2)
>>> jax[0].a # returns 4
```
cc: @zou3519

bdhirsh replied:
hmm I guess I would have thought that we would need some API like `pytree.register_leaf(DTensorSpec)`, instead of needing to update each call-site of `tree_map` (@zou3519 does that sound wrong?)

bdhirsh commented on torch/distributed/tensor/_api.py:
```diff
@@ -286,6 +286,11 @@ def __new__(
         return r
 
+    from torch._export.wrappers import _mark_constructor_exportable_experimental
+    @_mark_constructor_exportable_experimental
```
on the one hand, you could just put the decorator over DTensor's `__new__`, since they actually initialize their subclass contents in the __new__.

I guess adding an init here and putting it here is better for consistency though, in case people look at DTensor to copy code from

tugsbayasgalan replied:
I feel it is better to make the constructor work for just one case (either __new__ or __init__) because both have different semantics and concurrently supporting them feel bit fragile as the current logic is very specific to the implementaitons of them.

bdhirsh commented on torch/_export/passes/lift_constants_pass.py:
```diff
@@ -167,7 +167,10 @@ def lift_constants_pass(
             constant_val = _get_attr(gm, node.target)
-            if constant_val in lifted_objs:
+            if (
+                not isinstance(constant_val, torch.utils._pytree.TreeSpec)
```
are these changes specific to the subclass constructor support?

My guess is no - we'll also need these if export ever wants to support custom ops that accept arbitrary pytree types. Maybe worth a test for that as well? (cc @zou3519 )

tugsbayasgalan replied:
This is basically because now we have some new attrs on the graph module (input specs for flat_apply) and export tries to lift some constants. I refactored the code little bit to make it easier to understand.

bdhirsh commented on torch/_export/wrappers.py:
```diff
@@ -115,3 +126,52 @@ def call(self, *args):
+        # TODO container type input is not supported, will need to apply flat_apply
+        assert is_traceable_wrapper_subclass_type(type(args[0]))
+        constructor_subclass(*args, **kwargs)
+        if torch._C._is_torch_function_mode_enabled():
```
nit: invert the condition and return early instead of all the extr indenting (this will also make it clearer that this fn is meant to short circuit and be faster when we are not actively tracing pre-dispatch)

bdhirsh reviewed (commented):
high level looks good! Let me know when it's closer to landing

tugsbayasgalan commented on torch/_export/verifier.py:
```diff
@@ -206,7 +207,7 @@ def _allowed_op_types() -> tuple[type[Any], ...]:
 
             if not isinstance(op, _allowed_op_types()):
-                if op not in _allowed_builtin_ops() and op not in _allowed_torch_functions:
+                if op not in _allowed_builtin_ops() and op not in _allowed_torch_functions and not is_traceable_wrapper_subclass_type(op):
```
Ah i don't need it anymore i think

tugsbayasgalan committed:
Update on "[DRAFT] Support subclass constructor capturing in export"


Notable TODOs:
1. Need to implement AutogradHOP to get rid of subclasses before serializing
2. Need to implement mechanism to figure out what subclasses will be used in export when they are not expressed in the inputs 


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

Differential Revision: [D69640673](https://our.internmc.facebook.com/intern/diff/D69640673)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69640673).

bdhirsh commented on test/distributed/tensor/test_dtensor_compile.py:
```diff
@@ -127,6 +127,44 @@ def fn(x):
+                )
+                return inter.to_local()
+
+        torch.utils._pytree.register_constant(
```
why do we need this in addition to the new `is_leaf=lambda x: isinstance(x, DTensorSpec)` logic you added?

tugsbayasgalan replied:
This is so that flat_apply doesn't complain about it being non fx-able type.

pianpwk commented on torch/_export/wrappers.py:
```diff
@@ -115,3 +126,92 @@ def call(self, *args):
+                    constructor_subclass.__qualname__.lower().split(".")
+                )
+
+                # We actually don't want to create a new spec for each instance
```
I think we might actually want this (e.g. for dtensor, the spec contains tensor-specific info like strides, placement)

tugsbayasgalan replied:
sg!

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[DRAFT] Support subclass constructor capturing in export"


Notable TODOs:
1. Need to implement AutogradHOP to get rid of subclasses before serializing
2. Need to implement mechanism to figure out what subclasses will be used in export when they are not expressed in the inputs 


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

Differential Revision: [D69640673](https://our.internmc.facebook.com/intern/diff/D69640673)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69640673).

tugsbayasgalan committed:
Update on "[DRAFT] Support subclass constructor capturing in export"


Notable TODOs:
1. Need to implement AutogradHOP to get rid of subclasses before serializing
2. Need to implement mechanism to figure out what subclasses will be used in export when they are not expressed in the inputs 


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

Differential Revision: [D69640673](https://our.internmc.facebook.com/intern/diff/D69640673)

[ghstack-poisoned]

tugsbayasgalan committed:
Update

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[DRAFT] Support subclass constructor capturing in export"


Notable TODOs:
1. Need to implement AutogradHOP to get rid of subclasses before serializing
2. Need to implement mechanism to figure out what subclasses will be used in export when they are not expressed in the inputs 


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

Differential Revision: [D69640673](https://our.internmc.facebook.com/intern/diff/D69640673)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69640673).

tugsbayasgalan committed:
Update

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[DRAFT] Support subclass constructor capturing in export"


Notable TODOs:
1. Need to implement AutogradHOP to get rid of subclasses before serializing
2. Need to implement mechanism to figure out what subclasses will be used in export when they are not expressed in the inputs 


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

Differential Revision: [D69640673](https://our.internmc.facebook.com/intern/diff/D69640673)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69640673).

tugsbayasgalan committed:
Update

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "[DRAFT] Support subclass constructor capturing in export"


Notable TODOs:
1. Need to implement AutogradHOP to get rid of subclasses before serializing
2. Need to implement mechanism to figure out what subclasses will be used in export when they are not expressed in the inputs 


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

Differential Revision: [D69640673](https://our.internmc.facebook.com/intern/diff/D69640673)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69640673).

tugsbayasgalan committed:
Update on "[DRAFT] Support subclass constructor capturing in export"


Notable TODOs:
1. Need to implement AutogradHOP to get rid of subclasses before serializing
2. Need to implement mechanism to figure out what subclasses will be used in export when they are not expressed in the inputs 


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

Differential Revision: [D69640673](https://our.internmc.facebook.com/intern/diff/D69640673)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69640673).

bdhirsh commented on torch/testing/_internal/two_tensor.py:
```diff
@@ -36,6 +37,8 @@ def __new__(cls, a, b, outer_size=None, outer_stride=None):
         return out
 
+    @torch._disable_dynamo
+    @_mark_subclass_constructor_exportable_experimental
```
tbh this API is already in the private `torch._export` namespace, maybe we should just remove this leading underscore + experimental part of the string (otherwise you're going to have to change it at some point anyway, and if people start using it you'll still need to support both versions)

bdhirsh commented on torch/_export/wrappers.py:
```diff
@@ -115,3 +126,96 @@ def call(self, *args):
     return cls
+
+
+def _register_and_get_spec_proxy_in_tracer(tracer, name, spec):
```
can you explain what this is for

tugsbayasgalan replied:
Not needed! The context is i wanted to reuse the spec between different Dtensor init but i realized DeviceMesh etc can be different so it doesn't make sense to use same spec

tugsbayasgalan replied:
Actually we do need it for registering subclass constructor in fx graph and not duplicate it :)

bdhirsh replied:
hmm I guess it's not at all clear from the name / description of this fn that its main purpose is to take subclass constructor functions and put them as constant attributes on the graph. I guess it's theoretically more general than that, but it might be clearer to specialize the name etc just to make it clearer what the purpose of this function is?

Also - I think the expectation here is that nobody should be serializing/deserializing graphs with subclass constructors in it (at least, until we get the autograd HOP for subclasses working). Do you agree? And if so, can we add a loud error somewhere and a test for it, that way users get a good error message?

tugsbayasgalan replied:
Yep updated the name! Today, when people try to export a graph with subclass constructor and ser/der, it will error complaining it is not supported. This is because it will try to serialize Spec object directly. (see test_export.py expectedFailure)

bdhirsh commented on torch/_export/wrappers.py:
```diff
@@ -115,3 +126,96 @@ def call(self, *args):
+        constructor_subclass(*args, **kwargs)
+        if not torch._C._is_torch_function_mode_enabled():
+            return
+        torch_function_mode_stack = torch.overrides._get_current_function_mode_stack()
```
you still have a nested indent that you can fix here 😛 I would do something like:
```
        if not torch._C._is_torch_function_mode_enabled():
            return
       filtered_modes = [m for m in torch.overrides._get_current_function_mode_stack() if isinstance(m, PreDispatchTorchFunctionMode)]
       assert len(filtered_modes) <= 1, "duplicated pre-dispatch mode found"
       if len(filtered_modes) == 0:
           return
       predispatch_mode = filtered_modes[0]
       # regular logic...
```

bdhirsh commented on torch/_export/wrappers.py:
```diff
@@ -115,3 +126,96 @@ def call(self, *args):
+        ), "Can only apply _mark_constructor_exportable_experimental on tensor subclass constructors."
+        assert callable(constructor_subclass) and (
+            constructor_subclass.__name__ == "__init__"
+        ), "_mark_constructor_exportable_experimental can only be applied on tensor subclass constructors."  # noqa: B950
```
I think we can make this assert better - someone is inevitably going to apply it to something that is not their init function, and we want the error message to be actionable enough that they can fix it on their own

bdhirsh replied:
(1) you should move the assert **outside** of this `def wrapper`, so it can raise at startup time instead of needing to wait to raise until someone actually invokes the function

you also probably need a few levels of asserts to make this error message good. One outside of the wrapper (where we don't have any arguments yet), that just asserts that `constructor_subclass` is a function named "__init__", and its module corresponds to a type that subclasses `torch.Tensor`. Then you can have an even stronger assert inside of the `def wrapper` that asserts that `args[0]` is a tensor subclass, and `args[0].__init__ is constructor_subclass`

(2) you should make the error message clearly spit out (a) the name of the function that the user is incorrectly applying this decorator to, and (b) what they should apply it to instead (call out `{subclass_name}.__init__`)

I'm nitpicking here because I think we should do everything we can do make subclass errors more self serve!

bdhirsh commented on test/export/test_export.py:
```diff
@@ -11128,6 +11129,101 @@ def forward(self, x, y):
+    %access_subclass_inner_tensor_default_7 : [num_users=1] = call_function[target=torch.ops.export.access_subclass_inner_tensor.default](args = (%flat_apply, b), kwargs = {})
+    %twotensor___init__2 : [num_users=0] = get_attr[target=twotensor___init__0]
+    %twotensor___init__3 : [num_users=1] = get_attr[target=twotensor___init__2]
+    %twotensor_const_func_spec1 : [num_users=1] = get_attr[target=twotensor_const_func_spec1]
```
hmm... does this mean that we will have a different attribute on the graph for every `TwoTensor` constructor call?

It looks like the first thing you're doing in this PR is stashing the subclass constructor call itself as an attribute as the graph, because you can't pass it as argument (I think this makes sense!)

But don't we only need one attribute on the graph for each unique subclass type?

tugsbayasgalan committed:
Update on "Support subclass constructor capturing in export"


Notable TODOs:
1. Need to implement AutogradHOP to get rid of subclasses before serializing
2. Need to implement mechanism to figure out what subclasses will be used in export when they are not expressed in the inputs 


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

Differential Revision: [D69640673](https://our.internmc.facebook.com/intern/diff/D69640673)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69640673).

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support subclass constructor capturing in export"


Notable TODOs:
1. Need to implement AutogradHOP to get rid of subclasses before serializing
2. Need to implement mechanism to figure out what subclasses will be used in export when they are not expressed in the inputs 


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

Differential Revision: [D69640673](https://our.internmc.facebook.com/intern/diff/D69640673)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69640673).

bdhirsh commented on torch/distributed/tensor/_api.py:
```diff
@@ -282,6 +283,11 @@ def __new__(
+    @torch._disable_dynamo
+    @mark_subclass_constructor_exportable_experimental
+    def __init__(self, *args, **kwargs):
+        pass
```
this should probably be `super().__init__()` (otherwise if the default `torch.Tensor` constructor happens to have its own init, we won't be calling it at all)

bdhirsh reviewed (approved):
a few more comments but overall sgtm

tugsbayasgalan committed:
Update on "Support subclass constructor capturing in export"


Notable TODOs:
1. Need to implement AutogradHOP to get rid of subclasses before serializing
2. Need to implement mechanism to figure out what subclasses will be used in export when they are not expressed in the inputs 


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

Differential Revision: [D69640673](https://our.internmc.facebook.com/intern/diff/D69640673)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69640673).

tugsbayasgalan committed:
Update on "Support subclass constructor capturing in export"


Notable TODOs:
1. Need to implement AutogradHOP to get rid of subclasses before serializing
2. Need to implement mechanism to figure out what subclasses will be used in export when they are not expressed in the inputs 


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

Differential Revision: [D69640673](https://our.internmc.facebook.com/intern/diff/D69640673)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69640673).

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support subclass constructor capturing in export"


Notable TODOs:
1. Need to implement AutogradHOP to get rid of subclasses before serializing
2. Need to implement mechanism to figure out what subclasses will be used in export when they are not expressed in the inputs 


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

Differential Revision: [D69640673](https://our.internmc.facebook.com/intern/diff/D69640673)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support subclass constructor capturing in export"


Notable TODOs:
1. Need to implement AutogradHOP to get rid of subclasses before serializing
2. Need to implement mechanism to figure out what subclasses will be used in export when they are not expressed in the inputs 


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

Differential Revision: [D69640673](https://our.internmc.facebook.com/intern/diff/D69640673)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Support subclass constructor capturing in export"


Notable TODOs:
1. Need to implement AutogradHOP to get rid of subclasses before serializing
2. Need to implement mechanism to figure out what subclasses will be used in export when they are not expressed in the inputs 


cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

Differential Revision: [D69640673](https://our.internmc.facebook.com/intern/diff/D69640673)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69640673).
